Support Roberta embedding models

maxdebayser opened the pull request:
This PR adds support for Roberta embedding models. It's mostly the same as the Bert architecture, the only thing that changes is the padding token in the Embedding layer so this PR tries to reuse Bert modeling classes as much as possible. For some of the models we also need head size 32, so this size is added to the kernels here.

cc: @robertgshaw2-neuralmagic , @DarkLight1337 

FIX #9847

maxdebayser committed:
add support for Roberta models

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
support head size 32

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 commented:
Also, remember to update the Supported Models page.

DarkLight1337 commented:
It would also be nice to consider adding the `SequenceClassification` variant of the model (to solve #8022)

rpvelloso commented:
great! What's the ETA on this???

maxdebayser commented:
> Model implementation looks good (though a bit hacky). Can you add tests for the model?

Sure, I'll add the tests. I don't disagree that this is a bit hacky. Should we make the Bert classes more generic so that we can pass the embedding layer class as a parameter?

DarkLight1337 commented:
> > Model implementation looks good (though a bit hacky). Can you add tests for the model?
> 
> Sure, I'll add the tests. I don't disagree that this is a bit hacky. Should we make the Bert classes more generic so that we can pass the embedding layer class as a parameter?

That would be great. Another way would be to have an abstract `_init_embedding` etc. so subclasses can decide how to initialize each submodule.

maxdebayser committed:
fix after refactoring

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

flaviabeo committed:
Review suggestions

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

DarkLight1337 commented on vllm/model_executor/models/roberta.py:
```diff
@@ -0,0 +1,85 @@
+       _pooler: An instance of Pooler used for pooling operations.
+   """
+
+    def __init__(self,
+                 config: RobertaConfig,
+                 cache_config: Optional[CacheConfig] = None,
+                 quant_config: Optional[QuantizationConfig] = None,
+                 pooler_config: Optional[PoolerConfig] = None) -> None:
```
We should now use `VllmConfig` since #9983 . Also, please refactor the inner models as well.

flaviabeo committed:
Merge branch 'upstream_main' into roberta

flaviabeo committed:
Fixes conflicts with new upstream changes

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Merge changes fixes

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
More fixed related to the upstream merge

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Adds test for roberta model executor

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Asserts for Roberta models instance

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Fix space for linting

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Fix space for linting

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Modifies test for multilingual-e5-large

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

DarkLight1337 commented:
Please fix the formatting errors.

DarkLight1337 commented:
Also we have to merge from main to fix the CI failures.

flaviabeo committed:
Fix linting in test

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

flaviabeo committed:
Merge branch 'upstream_main' into roberta

DarkLight1337 reviewed (approved):
LGTM, thanks for adding this!

flaviabeo committed:
trigger ci

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

maxdebayser committed:
finish generalizing the Bert classes

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

flaviabeo committed:
Skips test for ROCm unsupported platform

Signed-off-by: Flavia Beo <flavia.beo@ibm.com>

maxdebayser committed:
fix roberta position_ids

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
@DarkLight1337, I realized that with Roberta models the position_ids start at padding_idx + 1 (see [here](https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/models/roberta/modeling_roberta.py#L133) and [here](https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/models/roberta/modeling_roberta.py#L1669) )

I've added a line of code to increment all position_ids by that amount. Without this, the results I get in the STS12 task from the MTEB benchmark for `intfloat/multilingual-e5-large` is this, 0.53, which is way off. With the change I get  0.80, which is correct. 

In my tests, all the position_ids in vllm for the embedding use case start with 0 and end with len()-1 and there are no padding tokens because the input tensors are 1-dimensional without padding. For example:

```
input_ids=tensor([     0,   1284,     70,   1821,   6275,    111,  21455,      6,      4,
            70,  21640,  31486,    111,     70,  15437,    509,     10,  63819,
             6,      5,      2,      0,   2367,     83,   1286,      6,      4,
            70,  16648,    111,  29700,    621,    959, 133888,     47, 137447,
          2363, 102880,    111,  77021,     47,  16839,  27289,      6,      5,
             2], device='cuda:0')
position_ids=tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
        15, 16, 17, 18, 19, 20, 21, 22, 23, 24], device='cuda:0')
```

Is there a scenario in which there could be a presence of padding tokens? (Except for the case in which the user inserts <pad> in the input text).

DarkLight1337 commented:
> I've added a line of code to increment all position_ids by that amount. Without this, the results I get in the STS12 task from the MTEB benchmark for intfloat/multilingual-e5-large is this, 0.53, which is way off. With the change I get 0.80, which is correct.

Hmm, we may need to add a correctness test that compares against HF then.

> Is there a scenario in which there could be a presence of padding tokens? (Except for the case in which the user inserts in the input text).

Don't think so, since vLLM encodes each prompt separately. Just to be sure, you can add an assertion statement so we know if our assumption is false.

sorenmc commented:
> It would also be nice to consider adding the `SequenceClassification` variant of the model (to solve #8022)

Are you still considering adding this to the pr? If not i could try to make an attempt.

maxdebayser committed:
add assert to verify assumption

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
@DarkLight1337 , I've added an `assert` on the `position_ids`.

> Hmm, we may need to add a correctness test that compares against HF then.

Yes, but it would have to be `sentence-transformers`. In the transformers library the pooled output is obtained by running the last hidden states through the `pooler` layer. But in sentence-transformers, this output is discarded and the hidden states are pooled using MEAN, CLS ... like we do and then normalized. Would it be OK to add this dependency? If yes, can we do this in another PR?

DarkLight1337 commented:
The existing embedding models tests for text-only models already use sentence-transformers, so it should be pretty straightforward to add this model to the list.

maxdebayser committed:
improve assert

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
add model to embedding test

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Remove encoder embedding model for compile test

The test is failing with Unsupported('dynamic shape operator: aten.nonzero.default; to enable, set torch._dynamo.config.capture_dynamic_output_shape_ops = True\n\nfrom user code:\n   File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/models/roberta.py", line 107, in forward\n    assert len(torch.nonzero(positions[start_pos])) == 0\n\nSet TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information\n\n\nYou can suppress this exception and fall back to eager by setting:\n    import torch._dynamo\n    torch._dynamo.config.suppress_errors = True\n')

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
trigger ci

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
trigger ci

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
trigger ci

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
The discussion on `SequenceClassification` models continues here in this other PR: https://github.com/vllm-project/vllm/pull/10400
